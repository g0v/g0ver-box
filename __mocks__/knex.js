import knex from 'jest-mock-knex';

export { client } from 'jest-mock-knex';

export default knex;

process.setMaxListeners(0);
