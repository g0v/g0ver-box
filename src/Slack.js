import _ from 'lodash';
import slack from 'slack';

export const bot = slack.rtm.client();

const TOKEN = process.env.SLACK_TOKEN;

if (TOKEN) {
  bot.listen({ token: TOKEN });
}

bot.hello((message) => {
  console.log(`Got a message: ${message.type}`);
});

const call = (name, req = {}) => (
  new Promise((resolve, reject) => {
    const fn = _.get(slack, name);
    fn(
      { ...req, token: TOKEN },
      (err, res) => (err ? reject(err) : resolve(res))
    );
  })
);

export default {
  userInfo: async req => await call('users.info', req),
  postMessage: async req => await call('chat.postMessage', { ...req, as_user: true }),
  postMultiMessage: async (users, message) => _.map(users, user => bot.ws.send(JSON.stringify({
    id: `${user}::${Date.now()}`, type: 'message', channel: user, text: message,
  }))),
  channelInfo: async req => await call('channels.info', req),
  channelJoin: async req => await call('channels.join', req),
};
