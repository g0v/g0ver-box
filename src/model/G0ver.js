import { isGlobalId, fromGlobalId, toGlobalId } from 'graphql-tower';
import CacheModel, { ValueColumn } from 'cache-model';
import db from './Database';
import GraphQLG0ver from '../type/GraphQLG0ver';

export default class User extends CacheModel {

  static tableName = 'g0ver';

  static database = db;

  static columns = {
    username: new ValueColumn(),
    skills: new ValueColumn(Object),
  }

  static fromGlobalId(id) {
    return isGlobalId(id) ? fromGlobalId(id, 'G0ver') : id;
  }

  static toGlobalId(id) {
    return toGlobalId('G0ver', id);
  }

  static toKeyword({ skills }) {
    return (skills || []).join(' ');
  }

  type = GraphQLG0ver;
}
