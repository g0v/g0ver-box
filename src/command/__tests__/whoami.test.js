import { client } from 'knex';
import Slack from '../../Slack';
import G0ver from '../../model/G0ver';
import index from '../';

describe('whoami command', () => {
  const data = { channel: 'D100', name: 'yutin', text: 'whoami' };

  it('when g0ver does not exist', async () => {
    client.mockReturnValueOnce([]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage).toHaveBeenLastCalledWith({ channel: 'D100', text: '請使用 `add <skill name>` 新增你的技能' });
  });

  it('when g0ver is existed', async () => {
    G0ver.prime('yutin');
    client.mockReturnValueOnce([{ id: 'yutin', skills: ['react', 'video'] }]);
    await index(data);
    expect(client).toMatchSnapshot();
    expect(Slack.postMessage).toHaveBeenLastCalledWith({ channel: 'D100', text: 'yutin 已登錄的技能 react, video' });
  });
});
