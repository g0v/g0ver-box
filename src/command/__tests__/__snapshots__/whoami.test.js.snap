// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whoami command when g0ver does not exist 1`] = `
Array [
  "select * from g0ver where id in (yutin) and deleted_at is null",
]
`;

exports[`whoami command when g0ver is existed 1`] = `
Array [
  "select * from g0ver where id in (yutin) and deleted_at is null",
]
`;
